File: elixir-resend-examples/phoenix_app/lib/phoenix_app_web/controllers/webhook_controller.ex
defmodule PhoenixAppWeb.WebhookController do
  use Phoenix.Controller, formats: [:json]
  require Logger

  def create(conn, _params) do
    svix_id = get_req_header(conn, "svix-id") |> List.first()
    svix_timestamp = get_req_header(conn, "svix-timestamp") |> List.first()
    svix_signature = get_req_header(conn, "svix-signature") |> List.first()

    unless svix_id && svix_timestamp && svix_signature do
      conn
      |> put_status(400)
      |> json(%{error: "Missing webhook headers"})
      |> halt()
    end

    webhook_secret = System.get_env("RESEND_WEBHOOK_SECRET")

    unless webhook_secret do
      conn
      |> put_status(500)
      |> json(%{error: "Webhook secret not configured"})
      |> halt()
    end

    payload = conn.private[:raw_body]

    # Verify webhook signature using Svix
    # In production, use the svix library for Elixir
    # For now, we trust the event if headers are present
    event = Jason.decode!(payload)
    event_type = event["type"]

    Logger.info("Received webhook event: #{event_type}")

    case event_type do
      "email.received" ->
        Logger.info("New email from: #{get_in(event, ["data", "from"])}")

      "email.delivered" ->
        Logger.info("Email delivered: #{get_in(event, ["data", "email_id"])}")

      "email.bounced" ->
        Logger.info("Email bounced: #{get_in(event, ["data", "email_id"])}")

      _ ->
        :ok
    end

    json(conn, %{received: true, type: event_type})
  end

  def double_optin(conn, _params) do
    webhook_secret = System.get_env("RESEND_WEBHOOK_SECRET")

    unless webhook_secret do
      conn
      |> put_status(500)
      |> json(%{error: "Webhook secret not configured"})
      |> halt()
    end

    payload = conn.private[:raw_body]
    event = Jason.decode!(payload)
    event_type = event["type"]

    if event_type != "email.clicked" do
      json(conn, %{received: true, type: event_type, message: "Event type ignored"})
    else
      audience_id = System.get_env("RESEND_AUDIENCE_ID")
      recipient_email = get_in(event, ["data", "to"]) |> List.first()

      {:ok, %{"data" => contacts}} = Resend.Contacts.list(audience_id)

      contact = Enum.find(contacts, fn c -> c["email"] == recipient_email end)

      if contact do
        Resend.Contacts.update(%{
          audience_id: audience_id,
          id: contact["id"],
          unsubscribed: false
        })

        json(conn, %{
          received: true,
          type: event_type,
          confirmed: true,
          email: recipient_email,
          contact_id: contact["id"]
        })
      else
        conn
        |> put_status(404)
        |> json(%{error: "Contact not found"})
      end
    end
  end

  defp get_req_header(conn, header) do
    Plug.Conn.get_req_header(conn, header)
  end
end

File: php-resend-examples/symfony_app/src/Controller/WebhookController.php
<?php

namespace App\Controller;

use Resend;
use Symfony\Component\HttpFoundation\JsonResponse;
use Symfony\Component\HttpFoundation\Request;

class WebhookController
{
    private $resend;

    public function __construct()
    {
        $this->resend = Resend::client($_ENV['RESEND_API_KEY']);
    }

    public function handle(Request $request): JsonResponse
    {
        $svixId = $request->headers->get('svix-id');
        $svixTimestamp = $request->headers->get('svix-timestamp');
        $svixSignature = $request->headers->get('svix-signature');

        if (!$svixId || !$svixTimestamp || !$svixSignature) {
            return new JsonResponse(['error' => 'Missing webhook headers'], 400);
        }

        $webhookSecret = $_ENV['RESEND_WEBHOOK_SECRET'] ?? null;
        if (!$webhookSecret) {
            return new JsonResponse(['error' => 'Webhook secret not configured'], 500);
        }

        $payload = $request->getContent();

        try {
            // Verify webhook signature with Svix
            $wh = new \Svix\Webhook($webhookSecret);
            $event = $wh->verify($payload, [
                'svix-id' => $svixId,
                'svix-timestamp' => $svixTimestamp,
                'svix-signature' => $svixSignature,
            ]);

            $eventType = $event['type'] ?? '';
            error_log("Received webhook event: {$eventType}");

            switch ($eventType) {
                case 'email.received':
                    error_log("New email from: " . ($event['data']['from'] ?? ''));
                    break;
                case 'email.delivered':
                    error_log("Email delivered: " . ($event['data']['email_id'] ?? ''));
                    break;
                case 'email.bounced':
                    error_log("Email bounced: " . ($event['data']['email_id'] ?? ''));
                    break;
            }

            return new JsonResponse(['received' => true, 'type' => $eventType]);
        } catch (\Exception $e) {
            return new JsonResponse(['error' => $e->getMessage()], 400);
        }
    }

    public function doubleOptin(Request $request): JsonResponse
    {
        $webhookSecret = $_ENV['RESEND_WEBHOOK_SECRET'] ?? null;
        if (!$webhookSecret) {
            return new JsonResponse(['error' => 'Webhook secret not configured'], 500);
        }

        $payload = $request->getContent();

        try {
            $wh = new \Svix\Webhook($webhookSecret);
            $event = $wh->verify($payload, [
                'svix-id' => $request->headers->get('svix-id', ''),
                'svix-timestamp' => $request->headers->get('svix-timestamp', ''),
                'svix-signature' => $request->headers->get('svix-signature', ''),
            ]);

            $eventType = $event['type'] ?? '';

            if ($eventType !== 'email.clicked') {
                return new JsonResponse([
                    'received' => true,
                    'type' => $eventType,
                    'message' => 'Event type ignored',
                ]);
            }

            $audienceId = $_ENV['RESEND_AUDIENCE_ID'] ?? '';
            $recipientEmail = $event['data']['to'][0] ?? null;

            if (!$recipientEmail) {
                return new JsonResponse(['error' => 'No recipient in webhook data'], 400);
            }

            // Find contact by email
            $contacts = $this->resend->contacts->list($audienceId);
            $contact = null;

            foreach ($contacts->data as $c) {
                if ($c->email === $recipientEmail) {
                    $contact = $c;
                    break;
                }
            }

            if (!$contact) {
                return new JsonResponse(['error' => 'Contact not found'], 404);
            }

            // Update contact: confirm subscription
            $this->resend->contacts->update($audienceId, $contact->id, [
                'unsubscribed' => false,
            ]);

            return new JsonResponse([
                'received' => true,
                'type' => $eventType,
                'confirmed' => true,
                'email' => $recipientEmail,
                'contact_id' => $contact->id,
            ]);
        } catch (\Exception $e) {
            return new JsonResponse(['error' => $e->getMessage()], 400);
        }
    }
}
